File: examples/cpp/linked_list_reverse.cpp
// Reverse a singly linked list. The VIZ_REGISTER_LINKED_LIST macro from
// viz.hpp tells the C++ tracer that `Node` is a linked-list node with
// fields `val` (the payload) and `next` (the recursive pointer). The
// tracer then emits `{kind: "object", type: "Node", fields: {val, next}}`
// for each node so the front-end renders the list with arrows.

#include <iostream>
#include "viz.hpp"

struct Node {
    int val;
    Node* next;
    Node(int v, Node* n = nullptr) : val(v), next(n) {}
};
VIZ_REGISTER_LINKED_LIST(Node, val, next)

Node* reverse(Node* head) {
    Node* prev = nullptr;
    Node* cur = head;
    while (cur != nullptr) {
        Node* nxt = cur->next;
        cur->next = prev;
        prev = cur;
        cur = nxt;
    }
    return prev;
}

int main() {
    Node* head = new Node(1, new Node(2, new Node(3, new Node(4))));
    Node* r = reverse(head);
    for (Node* p = r; p; p = p->next) std::cout << p->val << " ";
    std::cout << std::endl;
    return 0;
}

File: examples/python/linked_list_reverse.py
# Reverse a singly linked list in place.
# The visualizer will detect the `next` field and render this as a
# linked-list chain that rewires every iteration.

# @viz:Node=linked_list


class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next


def reverse(head):
    prev = None
    cur = head
    while cur is not None:
        nxt = cur.next   # save
        cur.next = prev  # rewire
        prev = cur       # advance prev
        cur = nxt        # advance cur
    return prev


def to_list(head):
    out = []
    while head is not None:
        out.append(head.val)
        head = head.next
    return out


if __name__ == "__main__":
    head = Node(1, Node(2, Node(3, Node(4, Node(5)))))
    new_head = reverse(head)
    print(to_list(new_head))
